src/container/srv_container.c

```diff
@@ -2547,3 +2551,163 @@ ds_cont_oid_fetch_add(uuid_t poh_uuid, uuid_t co_uuid, uuid_t coh_uuid,
+	if (rc != 0)
+		D_GOTO(out, rc);
+
+rechoose:
+	ep.ep_grp = NULL; /* primary group */
+	rc = rsvc_client_choose(&client, &ep);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": cannot find pool service: "DF_RC"\n",
+			DP_UUID(cont_uuid), DP_RC(rc));
+		D_GOTO(out_client, rc);
+	}
+
+	rc = cont_req_create(info->dmi_ctx, &ep, CONT_PROP_SET, &rpc);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to create cont set prop rpc: %d\n",
+			DP_UUID(cont_uuid), rc);
+		D_GOTO(out_client, rc);
+	}
+
+	in = crt_req_get(rpc);
+	uuid_clear(in->cpsi_op.ci_hdl);
+	uuid_clear(in->cpsi_op.ci_pool_hdl);
+	uuid_copy(in->cpsi_op.ci_uuid, cont_uuid);
+	uuid_copy(in->cpsi_pool_uuid, pool_uuid);
+	in->cpsi_prop = prop;
+
+	rc = dss_rpc_send(rpc);
+	out = crt_reply_get(rpc);
+	D_ASSERT(out != NULL);
+
+	rc = rsvc_client_complete_rpc(&client, &ep, rc,
+				      out->cpso_op.co_rc,
+				      &out->cpso_op.co_hint);
+	if (rc == RSVC_CLIENT_RECHOOSE) {
+		crt_req_decref(rpc);
+		dss_sleep(1000 /* ms */);
+		D_GOTO(rechoose, rc);
+	}
+
+	rc = out->cpso_op.co_rc;
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to set prop for container: %d\n",
+			DP_UUID(cont_uuid), rc);
+	}
+
+	crt_req_decref(rpc);
+out_client:
+	rsvc_client_fini(&client);
+out:
+	return rc;
+}
+
+void
+ds_cont_set_prop_handler(crt_rpc_t *rpc)
+{
+	int				 rc;
+	struct cont_svc			*svc;
+	struct cont_prop_set_in		*in = crt_reply_get(rpc);
+	struct cont_prop_set_out	*out = crt_reply_get(rpc);
+	struct rdb_tx			 tx;
+	uuid_t				 pool_uuid;
+	uuid_t				 cont_uuid;
+	struct ds_pool			*pool;
+	struct cont			*cont;
+	struct ds_cont_hdl		*hdl;
+	uuid_t				 hdl_uuid;
+
+	/* Client RPCs go through the regular flow with pool/cont handles */
+	if (daos_rpc_from_client(rpc)) {
+		ds_cont_op_handler(rpc);
+		return;
+	}
+
+	/*
+	 * Server RPCs don't have pool or container handles, so we need to fetch
+	 * the pool and container info.
+	 */
+	uuid_copy(pool_uuid, in->cpsi_pool_uuid);
+	uuid_copy(cont_uuid, in->cpsi_op.ci_uuid);
+
+	D_DEBUG(DF_DSMS, DF_CONT": processing cont set prop rpc %p\n",
+		DP_CONT(pool_uuid, cont_uuid), rpc);
+
+	rc = cont_svc_lookup_leader(pool_uuid, 0 /* id */,
+				    &svc, &out->cpso_op.co_hint);
+	if (rc != 0) {
+		D_ERROR("Failed to look up cont svc: %d\n", rc);
+		D_GOTO(out, rc);
+	}
+
+	/* If we have a container svc we must be on a pool svc node */
+	pool = ds_pool_lookup(pool_uuid);
+	if (pool == NULL) {
+		D_ERROR("Pool "DF_UUID" not found\n", DP_UUID(pool_uuid));
+		D_GOTO(out_svc, rc = -DER_NONEXIST);
+	}
+
+	rc = rdb_tx_begin(svc->cs_rsvc->s_db, svc->cs_rsvc->s_term, &tx);
+	if (rc != 0)
+		D_GOTO(out_pool, rc);
+
+	ABT_rwlock_wrlock(svc->cs_lock);
+
+	rc = cont_lookup(&tx, svc, cont_uuid, &cont);
+	if (rc != 0)
+		D_GOTO(out_lock, rc);
+
+	/* Coming from the server, set-prop is an administrative operation */
+	uuid_generate(hdl_uuid);
+	rc = ds_cont_local_open(pool_uuid, hdl_uuid, in->cpsi_op.ci_uuid, 0,
+				ds_sec_get_admin_cont_capabilities(), &hdl);
+	if (rc != 0)
+		D_GOTO(out_cont, rc);
```

Comment:
What is the purpose of this ds_cont_local_open call... I'm a bit concerned with the complex things it does. Are those necessary?

Reply:
Looks like you're already changing set_prop to take a capa parameter instead of a hdl.

Reply:
It looked like a handle UUID was required for the container IV call that set_prop does. Digging a few calls deeper, there's a comment on the cont_iv_key struct that's eventually used, saying it should be the container UUID for props, not the handle. Weird that this appeared to work to begin with if that's the case. Will experiment with this.

Reply:
Indeed, I don't understand why the container property IV needs a container handle... @wangdi1, any hints?

Reply:
Oh, it does not need container hdl, the initial IV interface is designed for capability, so using cont hdl uuid as the key. Hmm, the interface is a bit messy. I will cleanup it later.